src/compiler/checker.ts

```diff
@@ -6386,6 +6386,8 @@ namespace ts {
             // Only narrow when symbol is variable of type any or an object, union, or type parameter type
             if (node && symbol.flags & SymbolFlags.Variable) {
                 if (isTypeAny(type) || type.flags & (TypeFlags.ObjectType | TypeFlags.Union | TypeFlags.TypeParameter)) {
+                    const declaration = getDeclarationOfKind(symbol, SyntaxKind.VariableDeclaration);

```

Comment:
So now `getNarrowedTypeOfSymbol` walks up the parent chain twice, and further than it used to. Is there any performance impact?